spec/javascripts/packs/form-steps-wait.js

```diff
@@ -0,0 +1,116 @@
+import { fireEvent } from '@testing-library/dom';
+import { useSandbox } from '../support/sinon';
+import useDefineProperty from '../support/define-property';
+import { FormStepsWait } from '../../../app/javascript/packs/form-steps-wait';
+
+describe('FormStepsWait', () => {
+  const sandbox = useSandbox({ useFakeTimers: true });
+  const defineProperty = useDefineProperty();
+
+  function createForm({ action, method, options }) {
+    document.body.innerHTML = `
+      <form action="${action}" method="${method}" data-form-steps-wait="">
+        <input type="hidden" name="foo" value="bar">
+      </form>
+    `;
+
+    const form = document.body.firstElementChild;
+    Object.assign(form.dataset, options);
+    return form;
+  }
+
+  it('submits form via fetch', () => {
+    const action = new URL('/', window.location).toString();
+    const method = 'post';
+    const form = createForm({ action, method });
+    new FormStepsWait(form);
```

Comment:
seems weird to me to have a side-effecty constructor like this, what if we had a separate initializer like `FormStepsWait.createAndBind(form)` or something?

Reply:
> seems weird to me to have a side-effecty constructor like this, what if we had a separate initializer like `FormStepsWait.createAndBind(form)` or something?

Yeah, in retrospect, it's a good point. I think in my mind I was thinking of it in terms of being a common pattern for components to be fully-initialized via a mere constructor call. But as you've noted, this line in particular stands out as being especially "surprising" in obfuscating these details.

Your suggestion could be a good compromise, though in looking to some inspiration here ([uswds](https://github.com/uswds/uswds/blob/40f6a4257d276b77994a94be32cf105cd2417032/src/js/start.js#L21-L22), [other vanilla-JS component patterns](https://github.com/elix/elix#plain-html-and-javascript)), maybe it's fine enough to just have separate calls:

```js
const formStepsWait = new FormStepsWait(form);
formStepsWait.bind(/* target = document.body */);
```

Or, if we return `this` from the constructor, we could make it chainable:

```js
new FormStepsWait(form).bind();
```

Reply:
> Or, if we return `this` from the constructor, we could make it chainable:

You mean return `this` from `bind()`? (constructors always return the object no matter what?)

But yes, I like this suggestion a lot:

```js
new FormStepsWait(form).bind();
```

Reply:
> You mean return `this` from `bind()`? (constructors always return the object no matter what?)

No, I meant the constructor... I'd just overlooked the parenthesized fact 😅

Reply:
I still think I'll make the discussed revision, but wanted to capture an internal debate, since this felt "not super side-effecty" at least in how it's self-contained to / doesn't modify modify anything outside of the form itself. I think ultimately what I came to realize is that ideally we'd be _extending_ or _mixing into_ the base form, though there's not ([yet](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements#Customized_built-in_elements)) great primitives to be able to do this. Another option to express this could be something akin to `Array.from`, e.g. `FormStepsWait.from(form);`.

Reply:
> ```js
> new FormStepsWait(form).bind();
> ```

Updated in a4721ee.

Reply:
Yeah the idea of `FormStepsWait.from(form)` seems great, a much shorter name than the `FormStepsWait.createAndBind(form)` that I suggested

But yeah in terms of side effects, the thing that jumped out at me was using `new` and throwing away the result by not capturing it in a variable.